blocks/library/tag-cloud/block.js

```diff
@@ -0,0 +1,155 @@
+/**
+ * External dependencies
+ */
+import { map } from 'lodash';
+import { stringify } from 'querystringify';
+
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+import {
+	Placeholder,
+	ToggleControl,
+	SelectControl,
+	withAPIData,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import TermList from './term-list';
+import InspectorControls from '../../inspector-controls';
+import BlockControls from '../../block-controls';
+import BlockAlignmentToolbar from '../../block-alignment-toolbar';
+
+class TagCloudBlock extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.setTaxonomy = this.setTaxonomy.bind( this );
+		this.toggleShowTagCounts = this.toggleShowTagCounts.bind( this );
+	}
+
+	getTaxonomies() {
+		const taxonomies = this.props.taxonomies.data;
+
+		if ( ! taxonomies ) {
+			return [];
+		}
+
+		return map( taxonomies, taxonomy => {
+			return {
+				value: taxonomy.slug,
+				label: taxonomy.name,
+			};
+		} );
+	}
+
+	setTaxonomy( taxonomySlug ) {
+		const { setAttributes } = this.props;
+		const taxonomies = this.props.taxonomies.data;
+		const taxonomy = taxonomies[ taxonomySlug ];
+
+		setAttributes( { taxonomySlug, taxonomyRestBase: taxonomy.rest_base } );
+	}
+
+	toggleShowTagCounts() {
+		const { attributes, setAttributes } = this.props;
+		const { showTagCounts } = attributes;
+
+		setAttributes( { showTagCounts: ! showTagCounts } );
+	}
+
+	render() {
+		const {
+			attributes,
+			focus,
+			setAttributes,
+			termList,
+			className,
+		} = this.props;
+		const { taxonomySlug, showTagCounts, align } = attributes;
+		const taxonomies = this.getTaxonomies();
+		const options = [
+			{
+				label: __( '- Select -' ),
+				value: '',
+			},
+			...taxonomies,
+		];
+		const terms = termList.data ? (
+			<TermList
+				terms={ termList.data }
+				showTagCounts={ showTagCounts }
+				className={ className }
+			/>
+		) : (
+			[]
```

Comment:
Minor: It may be easier for React to reconcile† this in the render output if it were expressed as an empty value, e.g. `null` or `undefined`:

```js
let terms;
if ( termList.data ) {
	terms = (
		<TermList
			terms={ termList.data }
			showTagCounts={ showTagCounts }
			className={ className }
		/>
	);
}
```

Or:

```js
const terms = termList.data ? (
	<TermList
		terms={ termList.data }
		showTagCounts={ showTagCounts }
		className={ className }
	/>
) : null;
```

† The assumption being that it shortcuts to avoid iterating over an empty value.